### PR TITLE
[IMP] core: reduce registry memory footprint

### PR DESCRIPTION
**Pull request by rco-odoo**
This is a bunch of commits that reduce the memory taken by a registry.

On a registry with 296 modules, this saves 3 megabytes (out of 13.5) of memory, which is about 22% of the registry's memory footprint.

### CODE REVIEW

File: odoo/modules/registry.py [hunk 1]
```diff
@@ -135,6 +135,7 @@ def init(self, db_name):
         # field dependencies
         self.field_depends = Collector()
         self.field_depends_context = Collector()
+        self.field_inverses = Collector()
```
Comment (Elkasitu) [inline odoo/modules/registry.py:138]: good 👍🏽

File: odoo/modules/registry.py [hunk 2]
```diff
@@ -282,6 +284,8 @@ def setup_models(self, cr):
         for model in models:
             model._setup_complete()
 
+        del self._m2m
```
Comment (Elkasitu) [inline odoo/modules/registry.py:287]: I think we can define `self._m2m` right before `_setup_fields` and delete it right after, no?

**Review by Elkasitu — APPROVED** [review]

File: odoo/fields.py [hunk 3]
```diff
@@ -363,7 +363,7 @@ def _get_attrs(self, model_class, name):
         attrs['model_name'] = model_class._name
         attrs['name'] = name
         attrs['_module'] = modules[-1] if modules else None
-        attrs['_modules'] = set(modules)
+        attrs['_modules'] = tuple(unique(modules))
```
Comment (xmo-odoo) [inline odoo/fields.py:366]: `unique` creates a set internally, since order was clearly irrelevant why not just wrap the `tuple` call around the `set`?

**Review by xmo-odoo — COMMENTED** [review]
Possible alternative, though it would mostly be about sharing memory between registries (and might be harder to notice when profiling memory): `sys.intern` the related keys?

They're *field names*, which should already be interned (because they're symbol names), so all that should be needed is running `sys.intern` on them and blamo we'll retrieve the already interned strings and will only be paying for the tuple, which should commonly be cheaper than the joined string as it's a constant 40b + 8/item versus 49 + 1/char for `str`. The most common case for related being pairs the related tuple is 56b, which is enough for 6 characters (+ a `.`), which basically never happens.

On 3.8~3.9 anyway, 3.6 and 3.7 have less efficient tuples for some reason, respectively 48 and 56b baseline, meaning the cutoffs are at 14 and 22. I think we still have good odds of winning overall: out of ~1500 related, ~100 are under 14c and ~600 are under 22 (though these are rough numbers as I've not excluded related to sibling fields without periods, which should be counted differently as they'd get 8 less bytes worth of string).

File: odoo/tools/misc.py [hunk 4]
```diff
@@ -959,23 +959,26 @@ def update(self, *args, **kwargs):
...
+    __slots__ = ()
+
+    def __getitem__(self, key):
+        return self.get(key, ())
+
+    def __setitem__(self, key, val):
+        if val:
+            super().__setitem__(key, val)
+        else:
+            super().pop(key, None)
```
Comment (xmo-odoo) [inline odoo/tools/misc.py:986]: That seems pretty dodgy, and really doesn't match the docstring.

File: odoo/fields.py [hunk 5]
```diff
@@ -562,8 +562,8 @@ def setup_related(self, model):
             # add modules from delegate and target fields; the first one ensures
             # that inherited fields introduced via an abstract model (_inherits
             # being on the abstract model) are assigned an XML id
-            self._modules.update(model._fields[self.related[0]]._modules)
-            self._modules.update(field._modules)
+            delegate_field = model._fields[self.related[0]]
+            self._modules = tuple(unique(self._modules + delegate_field._modules + field._modules))
```
Comment (xmo-odoo) [inline odoo/fields.py:566]:
```suggestion
self._modules = tuple({*self._modules, *delegate_field._modules, *field._modules})
```
?

**Comment by rco-odoo** [comment]
@xmo-odoo about `field.related`: the commit is also about making the value type *consistent*.
Consider a related field defined as `related="foo_id.bar"`.  According to the new field setup (https://github.com/odoo/odoo/pull/69924), on its *definition class* the field will have `field.related = "foo_id.bar"`, while on its *registry class* it will have `field.related = ("foo_id", "bar")`.  Having that attribute as a tuple is not critical performance-wise.  And it feels more consistent to keep it as the developer put it.

**Comment by xmo-odoo** [comment]
> @xmo-odoo about `field.related`: the commit is also about making the value type _consistent_.
> Consider a related field defined as `related="foo_id.bar"`. According to the new field setup (#69924), on its _definition class_ the field will have `field.related = "foo_id.bar"`, while on its _registry class_ it will have `field.related = ("foo_id", "bar")`. Having that attribute as a tuple is not critical performance-wise. And it feels more consistent to keep it as the developer put it.

Sure, it's just a fair amount of change which we might be able to do without, and after thinking about interning I figured this change might even be the lesser memory optimisation. It's not like keeping the internal and external representations identical matters much.

File: odoo/tools/misc.py (hunk 4 above)
Comment (rco-odoo) [reply odoo/tools/misc.py:987]: I changed the docstring ;-)

**Comment by rco-odoo** [comment]
@robodoo rebase-ff r+

**Pull request closed without merging**